### PR TITLE
RBF Cross-Covariance

### PR DESCRIPTION
**Pull request by sanket-kamthe**
### Description
New  **feature**

- Adds RBF cross covariance in expectations. 
Previous implementation defaulted to quadrature expectations for cross co-variance when kernel and feature were not the same. 

- Should supersede previous pull request for the same feature  [#634](https://github.com/GPflow/GPflow/pull/634)
- Added test cases to cover all possible scenarios

- Cleaned up code based on [#672`](https://github.com/GPflow/GPflow/pull/672) which allowed 
```params_as_tensors_for()``` to take multiple arguments.

For e.g.,  long winding context manager
```python

 with params_as_tensors_for(kern1), params_as_tensors_for(feat1), \
        params_as_tensors_for(kern2), params_as_tensors_for(feat2):
````
is replaced by cleaner
```python
 with params_as_tensors_for(kern1, feat1, kern2, feat2):
````

### CODE REVIEW

File: tests/test_expectations.py [hunk 1]
```diff
@@ -101,6 +106,9 @@ def dirac_markov_gauss():
 def rbf_kern():
     return kernels.RBF(Data.D_in, variance=rng.rand(), lengthscales=rng.rand() + 1.)
 
+@cache_tensor
+def rbf_kern_2():
```
Comment (st--) [inline tests/test_expectations.py]: knowing what this PR is about I understand why you've got essentially the same code twice, but it'd be good to have a short comment explaining this duplication!

**Comment by st--** [comment]
@sanket-kamthe Looks like this is passing - anything else needed before we merge it?

**Comment by sanket-kamthe** [comment]
@st--  I will add the comment for the second rbf_kern() and then yes everything is ready I think for merge.

**Comment by awav** [comment]
@sanket-kamthe any updates on it?

File: gpflow/expectations.py [hunk 2]
```diff
@@ -230,7 +230,15 @@ def expectation(p, obj1, obj2=None, nghp=None):
     try:
         return _expectation(p, obj1, feat1, obj2, feat2, nghp=nghp)
     except NotImplementedError as e:
-        print(str(e))
+
+        def custom_warning_format(msg, *args, **kwargs):
```
Comment (awav) [inline gpflow/expectations.py]: I'm not sure about this lines. Do you still need this?

Comment (sanket-kamthe) [reply gpflow/expectations.py]: Yes, I would keep it. Otherwise the old msg was warning which was not informative at all. These two lines are the only way that I know of for format warning msg.

Also, unlike other print statements we eliminated, this is not a debug logger, we really want the user to know that specific expectation is solved by quadrature, if that is expected behaviour  then the user can choose to ignore the warning. 

I am not a fan of printing on stdout and logger is not appropriate for this case, as we are explicitly catching an error.

Comment (awav) [reply gpflow/expectations.py]: a lot of things have changed since you created this PR, we added logger and converted some warning to switchable logs

File: gpflow/expectations.py [hunk 3]
```diff
@@ -364,57 +372,90 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
     Compute the expectation:
     expectation[n] = <Ka_{Z1, x_n} Kb_{x_n, Z2}>_p(x_n)
         - Ka_{.,.}, Kb_{.,.} :: RBF kernels
-    Ka and Kb as well as Z1 and Z2 can differ from each other, but this is supported
-    only if the Gaussian p is Diagonal (p.cov NxD) and Ka, Kb have disjoint active_dims
-    in which case the joint expectations simplify into a product of expectations
+    Ka and Kb as well as Z1 and Z2 can differ from each other.
 
-    :return: NxMxM
+    :return: Nxdim(Z1)xdim(Z2)
```
Comment (awav) [inline gpflow/expectations.py]: N x dim(Z1) x dim(Z2)

File: gpflow/expectations.py [hunk 4]
```diff
@@ -364,57 +372,90 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
...
+            La, Lb = map(get_squared_length_scales, (Ka, Kb))
+            LaLb = La * Lb / (La + Lb)  # average length scale
 
-        sqrt_det_L = tf.reduce_prod(0.5 * squared_lengthscales) ** 0.5
-        C = tf.cholesky(0.5 * tf.matrix_diag(squared_lengthscales) + Xcov)  # NxDxD
+        sqrt_det_L = tf.reduce_prod(LaLb) ** 0.5
+        C = tf.cholesky(tf.matrix_diag(LaLb) + Xcov)  # NxDxD
         dets = sqrt_det_L / tf.exp(tf.reduce_sum(tf.log(tf.matrix_diag_part(C)), axis=1))  # N
 
+        def get_C_and_z_terms(Z, C=C):
```
Comment (awav) [inline gpflow/expectations.py]: docstrings: what are these C and z terms?

File: gpflow/expectations.py [hunk 5]
```diff
@@ -364,57 +372,90 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
...
+            z2_CC_inv_mu = z1_CC_inv_mu  # NxM
+            Z2 = Z1
+        else:
+            Z2, _ = Kb._slice(feat2.Z, p.mu)
+            C_inv_z2, z2_CC_inv_z2 = get_C_and_z_terms(Z2 / Lb * LaLb)
+            z2_CC_inv_mu = 2 * tf.matmul(C_inv_z2, C_inv_mu, transpose_a=True)[:, :, 0]  # NxM2
+
+        z1_CC_inv_z2 = tf.matmul(C_inv_z1, C_inv_z2, transpose_a=True)  # NxM1xM2
+
+        exponent_mahalanobis = mu_CC_inv_mu + tf.expand_dims(z2_CC_inv_z2, 1) + \
```
Comment (awav) [inline gpflow/expectations.py]: z2_CC_inv_z2 = tf.expand_dims(z2_CC_inv_z2, 1)
z1_CC_inv_z1 = tf.expand_dims(z1_CC_inv_z1, 2)
z1_CC_inv_mu = tf.expand_dims(z1_CC_inv_mu, 2) ...

File: gpflow/expectations.py [hunk 6]
```diff
@@ -364,57 +372,90 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
...
+        exponent_mahalanobis = tf.exp(-0.5 * exponent_mahalanobis)  # NxM1xM2
+
+        if Z1 == Z2:
+            # Compute sqrt(self.K(Z)) explicitly to prevent automatic gradient from
+            # being NaN sometimes, see pull request #615
+            dist_term = tf.exp(-0.25 * Ka.square_dist(Z1, None))
+        else:
+            # Compute exp( -.5 (Z-Z')^top (L_1+L_2)^{-1} (Z-Z') )
+            lengthscales_rms = tf.sqrt(La + Lb)
+            Z1 = Z1 / lengthscales_rms
```
Comment (awav) [inline gpflow/expectations.py]: `Z1 /= lengthscales_rms` and others

File: gpflow/expectations.py (hunk 4 above)
Comment (awav) [reply gpflow/expectations.py]: tbh, I think you need add some math clarification of what's going on here, _not only for this function_

File: tests/test_expectations.py (hunk 1 above)
Comment (awav) [reply tests/test_expectations.py]: @sanket-kamthe agreed, add docstring to this test describing what's going on

File: gpflow/expectations.py (hunk 2 above)
Comment (sanket-kamthe) [reply gpflow/expectations.py]: I know, but this is still valid case, as you can see original statement is still print.

Comment (awav) [reply gpflow/expectations.py]: point is that it is not really a scope of this task, It is global problem which is solvable by changing logger level or turning off warning messages

Comment (awav) [reply gpflow/expectations.py]: btw, `expectations.py` don't use warnings anymore, only logger

File: gpflow/expectations.py [hunk 7]
```diff
@@ -230,7 +230,15 @@ def expectation(p, obj1, obj2=None, nghp=None):
     try:
         return _expectation(p, obj1, feat1, obj2, feat2, nghp=nghp)
     except NotImplementedError as e:
-        print(str(e))
+
+        def custom_warning_format(msg, *args, **kwargs):
+            # ignore everything except the message
+            return str(msg) + '\n'
+
+        warnings.formatwarning = custom_warning_format
```
Comment (st--) [inline gpflow/expectations.py]: This needs to be reset to the previous version afterwards.

File: gpflow/expectations.py [hunk 8]
```diff
@@ -364,57 +372,95 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
...
-        squared_lengthscales = kern.lengthscales ** 2. if kern.ARD \
-            else tf.zeros((D,), dtype=settings.tf_float) + kern.lengthscales ** 2.
+        def get_squared_length_scales(kern):
+            squared_lengthscales = kern.lengthscales ** 2. if kern.ARD \
+                        else tf.zeros((D,), dtype=settings.tf_float) + kern.lengthscales ** 2.
+            return squared_lengthscales
 
-        sqrt_det_L = tf.reduce_prod(0.5 * squared_lengthscales) ** 0.5
-        C = tf.cholesky(0.5 * tf.matrix_diag(squared_lengthscales) + Xcov)  # NxDxD
+        if Ka == Kb:
```
Comment (st--) [inline gpflow/expectations.py]: For understanding correctly: does this mean "equal lengthscale parameters"?

File: gpflow/expectations.py [hunk 9]
```diff
@@ -364,57 +372,95 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
...
+            La = get_squared_length_scales(Ka)
+            Lb = La
+            LaLb = La * 0.5
+        else:
+            La, Lb = map(get_squared_length_scales, (Ka, Kb))
+            LaLb = La * Lb / (La + Lb)  # average length scale
+
+        sqrt_det_L = tf.reduce_prod(LaLb) ** 0.5
```
Comment (st--) [inline gpflow/expectations.py]: notation is slightly confusing/inconsistent with the maths - I understand you don't want to multiply and divide by 2, but then maybe better to call it half_geom_mean_scale (or something along those lines)?

**Comment by st--** [comment]
This PR needs to undo the deprecation of Stationary.scaled_square_dist() in kernels.py

File: gpflow/expectations.py (hunk 8 above)
Comment (sanket-kamthe) [reply gpflow/expectations.py]: Yes, equal length scales, we use variances at the end only.

File: gpflow/expectations.py (hunk 9 above)
Comment (sanket-kamthe) [reply gpflow/expectations.py]: calling it  `half_mean_L`

File: gpflow/expectations.py [hunk 10]
```diff
@@ -110,9 +110,6 @@ def _quadrature_expectation(p, obj1, feature1, obj2, feature2, num_gauss_hermite
     """
     num_gauss_hermite_points = 100 if num_gauss_hermite_points is None else num_gauss_hermite_points
 
-    logger.warn("Quadrature is used to calculate the expectation. This means that "
-                "an analytical implementations is not available for the given combination.")
```
Comment (st--) [inline gpflow/expectations.py]: implementations -> implementation

File: gpflow/expectations.py [hunk 11]
```diff
@@ -364,57 +361,108 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
...
+        # along M2
+        z1_CC_inv_mu = tf.expand_dims(z1_CC_inv_mu, 2)
+        z1_CC_inv_z1 = tf.expand_dims(z1_CC_inv_z1, 2)
+
+        # expanded version of ((Z1 + Z2)-mu) (CCT)-1 ((Z1 + Z2)-mu)
+        exponent_mahalanobis = mu_CC_inv_mu + z2_CC_inv_z2 + \
+                               z1_CC_inv_z1 + 2 * z1_CC_inv_z2 - \
+                               z1_CC_inv_mu- z2_CC_inv_mu  # NxM1xM2
+
+        exponent_mahalanobis = tf.exp(-0.5 * exponent_mahalanobis)  # NxM1xM2
```
Comment (rhaps0dy) [inline gpflow/expectations.py]: I think it would make sense to name this left-hand-side `mahalanobis`

File: gpflow/expectations.py [hunk 12]
```diff
@@ -364,57 +361,108 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
...
+
+        if Z1 == Z2:
+            # CAVEAT : Compute sqrt(self.K(Z)) explicitly
+            # to prevent automatic gradient from
+            # being NaN sometimes, see pull request #615
+            dist_term = tf.exp(-0.25 * Ka.scaled_square_dist(Z1, None))
```
Comment (rhaps0dy) [inline gpflow/expectations.py]: Even if `Z1 == Z2`, shouldn't the lengthscales here be `lengthscales_rms` like in the `else` branch? Hold on, I'm going to build a failing test case.

Comment (sanket-kamthe) [reply gpflow/expectations.py]: I don't think so, as this Z1 and Z2 are not same as feat1.Z and feat2.Z they are already scaled. Moreover, unless the objects are not same this branch is never visited. This condition was added to make sure we do not lose performance for both rbfs being exactly same.

Comment (ialong) [reply gpflow/expectations.py]: I agree with @sanket-kamthe here. `Ka.scaled_square_dist` takes care of the square rooting and scaling.

**Comment by st--** [comment]
@sanket-kamthe Hi Sanket, would you be up for porting this PR to gpflow 2 ?
